libs/libcds/tests/unit/libcds_feldman_map_hazard_pointer.cpp

```diff
@@ -23,39 +25,96 @@
+    {
+    }
+
+    atomwrapper(const atomwrapper& other)
+      : _a(other._a.load())
+    {
+    }
+
+    atomwrapper& operator=(const atomwrapper& other)
+    {
+        _a.store(other._a.load());
+    }
+};
+
 using gc_type = cds::gc::custom_HP<cds::gc::hp::details::HPXTLSManager>;
 using key_type = std::size_t;
-using value_type = std::string;
+using value_type = atomwrapper<std::size_t>;
 
 template <typename Map>
-void run(Map& map, const std::size_t nMaxItemCount)
+void run(Map& map, const std::size_t n_items, const std::size_t n_threads)
 {
-    std::vector<key_type> rand_vec(nMaxItemCount);
-    std::generate(rand_vec.begin(), rand_vec.end(), std::rand);
+    // a reference counter vector to keep track of counter on
+    // value [0, 1000)
+    std::array<std::atomic<std::size_t>, 1000> counter_vec;
+    std::fill(std::begin(counter_vec), std::end(counter_vec), 0);
+
+    std::vector<hpx::thread> threads;
 
-    std::vector<hpx::future<void>> futures;
+    // each thread inserts number of n_items/n_threads items to the map
+    std::vector<std::vector<std::size_t>> numbers_vec(
+        n_threads, std::vector<std::size_t>(n_items / n_threads, 0));
 
-    for (auto ele : rand_vec)
+    // map init
+    std::size_t val = 0;
+    while (val < 1000)
     {
-        futures.push_back(hpx::async([&, ele]() {
-            // enable this thread/task to run using libcds support
-            hpx::cds::hpxthread_manager_wrapper cds_hpx_wrap;
+        std::atomic<std::size_t> counter(0);
+        map.insert(val, counter);
+        val++;
+    }
 
+    auto insert_val_and_increase_counter =
+        [&](std::vector<std::size_t>& number_vec) {
+            hpx::cds::hpxthread_manager_wrapper cds_hpx_wrap;
             hpx::this_thread::sleep_for(std::chrono::seconds(rand() % 5));
-            map.insert(ele, std::to_string(ele));
-        }));
-    }
+            for (auto val : number_vec)
+            {
+                typename Map::guarded_ptr gp;
+
+                gp = map.get(val);
+                if (gp)
+                {
+                    (gp->second)._a++;
+                    counter_vec[gp->first]++;
+                }
+            }
+        };
+
+    for (auto& v : numbers_vec)
+    {
+        std::generate(v.begin(), v.end(),
+            [&]() { return rand() % (n_items / n_threads); });
 
-    hpx::wait_all(futures);
+        threads.emplace_back(insert_val_and_increase_counter, std::ref(v));
+    }
 
-    std::size_t count = 0;
+    // wait for all threads to complete
+    for (auto& t : threads)
+    {
+        if (t.joinable())
+            t.join();
+    }
 
-    while (!map.empty())
+    for (auto it = map.cbegin(); it != map.cend(); ++it)
     {
-        auto guarded_ptr = map.extract(rand_vec[count]);
-        HPX_ASSERT(guarded_ptr->first == rand_vec[count]);
-        HPX_ASSERT(guarded_ptr->second == std::to_string(rand_vec[count]));
-        count++;
+        HPX_TEST_EQ(counter_vec[it->first], (it->second)._a);
```

Comment:
This is much more like what I expected. Now the maps are tested more fully. 
One thing that confuses me - the sleep for random seconds - this happens once on task create and if I recall correctly, we wanted a sleep in the tasks to allow them to suspend and resume - possibly on a different kernel thread - this makes sure that the test covers the possibility that thread local storage is moving from one core to another correctly. However, putting a single sleep at the start of each task is not very useful.
Suggestion. Move the sleep into the main loop so that instead of this
```
    auto insert_val_and_increase_counter =
        [&](std::vector<std::size_t>& number_vec) {
            hpx::cds::hpxthread_manager_wrapper cds_hpx_wrap;
            hpx::this_thread::sleep_for(std::chrono::seconds(rand() % 5));
            for (auto val : number_vec)
            {
                typename Map::guarded_ptr gp; 

                gp = map.get(val);
                if (gp)
                {
                    (gp->second)._a++;
                    counter_vec[gp->first]++;
                }
            }                                                                                                                                                                                 
        };
```
we have   
```
    auto insert_val_and_increase_counter =
        [&](std::vector<std::size_t>& number_vec) {
            hpx::cds::hpxthread_manager_wrapper cds_hpx_wrap;
            for (auto val : number_vec)
            {
                if (rand() %10 == 0) hpx::this_thread::yield();
                typename Map::guarded_ptr gp;
                gp = map.get(val);
                if (gp)
                {
                    (gp->second)._a++;
                    counter_vec[gp->first]++;
                }
            }                                                                                                                                                                                 
        };
```
why? because now the task will suspend itself many times during the insertion phase and all the threads that are active will be hitting the map in random ordering as they suspend and resume constantly. This will really make sure that if any thread local data is corrupted, then we'll know about it.
I am not familiar with the `guarded_ptr`- should the `if` block also have an `else throw(std::runtime_error(...));` just to make sure we know if something fails? (I think that the map is initialized with all possible `val` values on start.